### ISSUE TITLE
Removing subheaders content identifiers

### ISSUE THREAD

**Issue by ioggstream** [httpwg/http-extensions#1315]
## Discussion

At the last http-wg interim meeting we discussed the opportunity to have sub-headers content identifiers, and it was noted that they are subject to a variety of problems.

I suggest to:

- remove them in the current draft
- eventually refining the proposal in a new PR with the support of the ML

I would be great to get more information about the use cases of this features and share some thoughts with the wg.

**Comment by msporny** [comment]
Can you please explain what happened in the HTTP WG interim meeting for those of us that could not attend? Exactly what is the problem and could you provide a concrete example of the problem?

**Comment by ioggstream** [comment]
Hi @msporny, hope you're fine :)

Here are the [interim minutes](https://httpwg.org/wg-materials/interim-20-10/minutes.html). I don't know if there are some recordings.

One of the problem is the following: 

1- if Signature references a sub-header, the verifier must know the syntax of the header
2- the header will thus be parsed before validating the signature
3- it does not require to send a valid signature to force the verifier to parse (split, token, indexing, ...) whatever header
4- the verifier might not know the header syntax: at this point it can either guess or fail
5- if they guess, the sender can steer the way the signature is processed

@martinthomson  explained better than me a generalization of this issue.

**Comment by richanna** [comment]
> 1- if Signature references a sub-header, the verifier must know the syntax of the header
> 2- the header will thus be parsed before validating the signature

The verifier only needs to know how to parse a Structured Field. It doesn't need to know anything about what the contents mean, or what is or isn't actually permitted by the field's specification. It does not need to interpret the field's value, or act on it in any way.

> 3- it does not require to send a valid signature to force the verifier to parse (split, token, indexing, ...) whatever header

I'm not seeing the concern here. The only thing I can think of is that a malicious signer could try to blow up the verifier by signing a single key out of an absurdly large header field value. But that's a vulnerability in the Structured Field parser implementation, not this draft.

> 4- the verifier might not know the header syntax: at this point it can either guess or fail
> 5- if they guess, the sender can steer the way the signature is processed

See above; the verifier does not need to know anything about the header field, it only needs to know how to parse Structured Fields. We can make it more explicit that a signer is only permitted to use these with Structured Fields, but even if they used it with a non-SF header field, it might not provide the protection they want, but it wouldn't pose an availability risk to the verifier.

**Comment by ioggstream** [comment]
> The verifier only needs to know how to parse a Structured Field

Not all fields are S-F.

> a malicious signer ..

iiuc you don't need a malicious signer, just an intermediary: this is because the signature should be validated only after headers are processed and the issue triggered.

That's my 2¢ (I'm not a security expert :P) but I think the concerns raised during the interim were grounded.

**Comment by richanna** [comment]
>> The verifier only needs to know how to parse a Structured Field
>
> Not all fields are S-F.

The dictionary structured field member and list prefix content identifiers by definition identify contents within Structured Fields and cannot be used with unstructured fields.

See section 2.2:
> An individual member in the value of a Dictionary Structured Field is identified by...

And section 2.3:
> A prefix of a List Structured Field consisting of...

We can add an explicit "MUST NOT use these with unstructured fields" message, to make this clearer.

> iiuc you don't need a malicious signer, just an intermediary: this is because the signature should be validated only after headers are processed and the issue triggered.

If the message isn't sent over TLS, or if the intermediary terminates the TLS connection, then sure. That risk exists *now*, if the intermediary can inject a malicious value into a structured field that the recipient is expected to process. Given that the `Signature` and `Signature-Input` header fields themselves are structured fields, the intermediary could just as easily inject their malicious value into one of those fields, and blow up the recipient when it tries to process the message signature itself. We could try to defend against this by imposing strict limitations on these header fields, but this would limit the extensibility of the spec, and we'd be guessing at what those limitations should be.

All of this underscores the fact that this is a Structured Field problem, not a Message Signatures problem. This risk is not unique to Message Signatures, and cannot be effectively mitigated within Message Signatures (or any other structured field definition for that matter). HTTP libraries need to defend against malicious structured field values, just as they have to defend against maliciously crafted HTTP messages.

**Comment by jricher** [comment]
This situation should hopefully be more clear as of -03 and -04, and the core draft no longer relies on subfield indexing for core functionality. As this is a structured field issue, I think the best approach is to add a security consideration.

**Comment by jricher** [comment]
List prefixes have been removed by #1529 and sufficient use for keeping dictionary key access exists for multiple signatures. Closing this issue.
